lib/TPP/CombineXsmmPass.cpp

```diff
@@ -0,0 +1,179 @@
+namespace tpp {
+#define GEN_PASS_DEF_COMBINEXSMMOPPASS
+#include "TPP/Passes.h.inc"
+} // namespace tpp
+} // namespace mlir
+namespace {
+
+static FailureOr<DenseI64ArrayAttr>
+getSizesAndLeadingDimForBrgemmOp(RewriterBase &rewriter, xsmm::BrgemmOp opTy) {
+
+  auto memrefC = opTy.getOperand(3).getType();
+  auto memrefA = opTy.getOperand(1).getType();
+  auto memrefB = opTy.getOperand(2).getType();
+
+  int64_t m = memrefC.cast<ShapedType>().getShape()[0];
+  int64_t n = memrefC.cast<ShapedType>().getShape()[1];
+  int64_t k = memrefA.cast<ShapedType>().getShape()[2];
+
+  auto ldaDim = xsmm::utils::getLeadingDim(memrefA, /*pos=*/1);
+  if (failed(ldaDim)) {
+    return failure();
+  }
+  int64_t lda = *ldaDim;
+
+  auto ldbDim = xsmm::utils::getLeadingDim(memrefB, /*pos=*/1);
+  if (failed(ldbDim)) {
+    return failure();
+  }
+  int64_t ldb = (vnni::utils::isInVnniLayout(memrefB.cast<MemRefType>()))
+                    ? *ldbDim / *vnni::utils::getVnniBlockingFactor(memrefB)
+                    : *ldbDim;
+
+  auto ldcDim = xsmm::utils::getLeadingDim(memrefC);
+  if (failed(ldcDim)) {
+    return failure();
+  }
+  int64_t ldc = *ldcDim;
+
+  // If we are dealing with a BRGEMM we need to pass two extra dimensions:
+  // - strideA and strideB that represent the stride between different GEMM
+  // in BRGEMM.
+  int64_t strideA = lda * m;
+  int64_t strideB = ldb * k;
+  return DenseI64ArrayAttr::get(
+      rewriter.getContext(),
+      ArrayRef<int64_t>{m, n, k, lda, ldb, ldc, strideA, strideB});
+}
+
+static ArrayAttr getBrgemmFlags(RewriterBase &rewriter, xsmm::BrgemmOp opTy) {
+  auto memrefB = opTy.getOperand(2).getType().cast<MemRefType>();
+  xsmm::GemmFlagsAttr gemmFlag =
+      (vnni::utils::isInVnniLayout(memrefB))
+          ? xsmm::GemmFlagsAttr::get(rewriter.getContext(),
+                                     xsmm::GemmFlags::VNNI_B)
+          : xsmm::GemmFlagsAttr::get(rewriter.getContext(),
+                                     xsmm::GemmFlags::NONE);
+  return rewriter.getArrayAttr(gemmFlag);
+}
+
+struct CombineXsmmOp : public OpRewritePattern<xsmm::BrgemmOp> {
+  using OpRewritePattern<xsmm::BrgemmOp>::OpRewritePattern;
+  LogicalResult matchAndRewrite(xsmm::BrgemmOp brgemmOp,
+                                PatternRewriter &rewriter) const override {
+    auto brgemmResultOperand = brgemmOp.getOperand(3);
+    for (auto use = brgemmResultOperand.user_begin();
+         use != brgemmResultOperand.user_end(); use++) {
+      auto user = *use;
+      if (isa<xsmm::BinaryOp>(user) &&
+          (dyn_cast<xsmm::BinaryOp>(user)).getCallee() ==
+              mlir::xsmm::BinaryKind::ADD) {
+        xsmm::BinaryOp binaryOp = ((xsmm::BinaryOp)user);
+        auto dispatchOp = dyn_cast<xsmm::BinaryDispatchOp>(
+            binaryOp.getOperand(0).getDefiningOp());
+        auto isFusedAdd = dispatchOp.getKind() == xsmm::BinaryKind::ADD;
+        auto binaryFlags = dispatchOp.getFlags();
+        if (isFusedAdd &&
```

Comment:
Could this be a matcher? @chelini

Reply:
No, current matchers allow you to check the properties of a single op; here, you need something to detect a DAG. MLIR has some matchers for this: https://github.com/llvm/llvm-project/blob/main/mlir/include/mlir/IR/Matchers.h